# <font color='blue'> Ebay Car Sales - Numpy & Pandas </font>  

The purpose of this project is to clean a dataset.  The data is taken from eBay's used car classified ads from a German eBay website, eBay Kleinanzeigen.  The original dataset was scraped and uploaed to Kaggle by user orgesleka.  Although, the data is no longer found on Kaggle, it can be downloaded from here:  https://data.world/data-society/used-cars-data

The dataset contains 50,000 rows of data, and information on the vehicle, seller, date when the ad was first crawled, and when the crawler last saw the ad online.  

**Read in the dataset, autos, and import Numpy and pandas.**

In [1]:
# Read in the csv file, autos, as a dataframe.  And import Numpy and pandas.

import numpy as np
import pandas as pd
#encoding can be "Latin-1" or "windows-1252"
autos = pd.read_csv('C:/Users/Name/Documents/PythonScripts/DataSets/autos.csv', encoding="Latin-1")

**DATA EXPLORATION**

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The columns, price and odometer, are object type that need to be numeric type.

In [3]:
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

**DATA CLEANING: Column Names**

To make the column names more readable, make the changes:
- Change column headers from camelcase to snakecase. 
- Reword some of the column headers to make them more descriptive. 

In [4]:
# View column headers before changes.
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
# Change column header names
autos.columns = ["date_crawled", "name","seller","offer_type","price","ab_test",
"vehicle_type","registration_year","gearbox","power_ps",
"model","odometer","registration_month","fuel_type","brand",
"unrepaired_damage","ad_created","num_pictures",
"postal_code","last_seen"]

# View column headers after changes.
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

There are a lot of columns, and some of them may be unnecessary.  If a column only has a few unique values, then that indicates that most of the values are the same and the column is a candidate for removal.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

The columns, seller, offerType, and notRepairedDamage have a low number of unique values.  The notRepairedDamage column will be investigated to determine if it should be dropped.

**DATA EXPLORATION: unrepaired_damage column**

In [7]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

The notRepairedDamage column has only 2 unique values, "nein" or no, and "ja" or yes.  This could represent either in repair or not in repair, so this column will not be removed.

**DATA EXPLORATION: price and odometer column**

In [8]:
# Price and odometer column: view first rows

print(autos["price"].head())
print('\n')
print(autos["odometer"].head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object


0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


**DATA CLEANING: price and odometer columns**

Some data cleaning of the price and odometer columns is needed:

- Price column:
    - remove "$" and ","  
    - convert to integer type


- Odometer column:
    - remove "," and "km"
    - convert to integer type
    - rename the column to odometer_km

In [9]:
# Clean the odometer column.

autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

In [10]:
# Clean the price column.

autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

**DATA CLEANING: remove outliers in price and odometer columns.**

The range for the price and odometer columns may be unreasonable.  Outliers will be considered outside of the 25% and 75% percentile.  The next section will look for outliers in the price and odometer columns.

In [11]:
# View outiers in the odometer_km column.

print("'odometer_km' COLUMN - NUMBER OF UNIQUE VALUES: ", 
      autos["odometer_km"].unique().shape)
print('\n')
print("'odometer_km' COLUMN - SOME STATISTICS: ")
print(autos["odometer_km"].describe())
print('\n')
print("'odometer_km' COLUMN - INDEX NUMBER & VALUE: ")
autos["odometer_km"].value_counts().head()

'odometer_km' COLUMN - NUMBER OF UNIQUE VALUES:  (13,)


'odometer_km' COLUMN - SOME STATISTICS: 
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


'odometer_km' COLUMN - INDEX NUMBER & VALUE: 


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

The minimum in the odometer_km column is 5000, which is a possible minimum odometer reading, so values below 25% will be retained.  The 75% percentile is the same as the maximum value, so there are no outliers in the odometer_km column.  

In [12]:
# View outiers in the price column.

print("'price' COLUMN - NUMBER OF UNIQUE VALUES: ", 
      autos["price"].unique().shape)
print('\n')
print("'price' COLUMN - SOME STATISTICS: ")
print(autos["price"].describe())
print('\n')
print("'price' COLUMN - INDEX NUMBER & VALUE: ")
autos["price"].value_counts().head()

'price' COLUMN - NUMBER OF UNIQUE VALUES:  (2357,)


'price' COLUMN - SOME STATISTICS: 
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


'price' COLUMN - INDEX NUMBER & VALUE: 


0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

Values below the 25th percentile may cover low bids, so below the 25th percentile will not be considered outliers.  There is a significant jump from the 75th percentile (10^3) to the maximum value (10^8), so values above the 75th percentile will be explored further.

In [13]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

For the price column, the values above the 75th percentile appear to be outliers and will be removed.  

In [14]:
# Remove outliers > 75th percentile in the price column.

autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Outliers in the price column have been removed.  The minimum is 1 and the maximum is 350,000, which is a more reasonable range.

**DATA EXPLORATION of the date columns: date_crawled, last_seen, ad_created, registration_month, registration_year.**

Two date columns were created by the crawler, date_crawled, last_seen.  The other three columns were created by the website, ad_created, registration_month, registration_year.

In order to evaluate the dates quantitatively, they would need to be converted to numeric type.

In [15]:
# String Type Date columns: view date formatting for string type columns

autos[['date_crawled','ad_created','last_seen']][0:3]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37

In [16]:
# date_crawled column: view just the date, or the first ten characters of the string.

print(autos['date_crawled'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object


For string columns, date_crawled, last_seen, ad_created: create frequency distribution of dates.

In [17]:
# Create frequency distribution of dates for date_crawled column.

print("date_crawled Column - DATE....FREQUENCY:")
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head(20)

date_crawled Column - DATE....FREQUENCY:


2016-04-07    0.001400
2016-04-06    0.003171
2016-04-05    0.013096
2016-04-04    0.036487
2016-04-03    0.038608
2016-04-02    0.035478
2016-04-01    0.033687
2016-03-31    0.031834
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-28    0.034860
2016-03-27    0.031092
2016-03-26    0.032204
2016-03-25    0.031607
2016-03-24    0.029342
2016-03-23    0.032225
2016-03-22    0.032987
2016-03-21    0.037373
2016-03-20    0.037887
2016-03-19    0.034778
Name: date_crawled, dtype: float64

For date_crawled column, it appears that the site was crawled for about a month from March 19 through April 7 in 2016. The distribution of listings on each day is roughly the same.

In [18]:
# Create frequency distribution of dates for last_seen column.

autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last_seen column, is the date the crawler last saw any listings, which is likely the date the listing was removed, perhaps due to the car being sold.

In the last few days, there was a disproportionate amount of last seen dates. This may possibly due to the crawling period ending, rather than a huge number of car sales.

In [19]:
# Create frequency distribution of dates for ad_created column.

autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

For ad_created column, there's a range in the dates where ads were created. Low frequency of ads were created in 2015 and early part of 2016. Most of the ads were created in March and April of 2016.

**DATA EXPLORATION:  registration_year column**

In [20]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration_year is likely the year the vehicle was first purchased and corresponds to the age of the car. The minimum, 1000, and maximum years, 9999, are nonsensical. Year 1000 is before the time when vehicles were manufactured, and year 9999 is in the future. Also, registering a car after the ad was created doesn't make any sense.

A reasonable date range may be 1900 - 2016.  The dates before 1900 will be investigated.

In [21]:
# Find the percent of dates prior to 1900.

reg_yr_after1900 = (autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]
reg_yr_before1900 = round((100 - reg_yr_after1900 * 100), 1)


print('Percent of vehicles registered before 1900:  ')
print(reg_yr_before1900)

Percent of vehicles registered before 1900:  
3.9


Only 4% of the dates are prior to 1900, which is a reasonable reason to remove the dates prior to 1900.

**DATA CLEANING - registration_year: remove dates prior to 1900.**

In [22]:
# View the number of each date prior to 1900.

autos['registration_year'].value_counts(dropna=False).sort_index().head()

1000    1
1001    1
1111    1
1800    2
1910    5
Name: registration_year, dtype: int64

In addition, the date prior to 1910 is 1800, and vehicles were not manufactured by that date. So the earlier hypothesis for the date range is reasonable, 1900 to 2016.

In [23]:
# Registration_year column: set the registration year between 1900 and 2016.

autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [24]:
# Registration_year column: verify date range.

autos['registration_year'].value_counts(normalize=True).sort_index(ascending=True)

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

From the registration_year column, it looks like most of the vehicles were registered from 1994 - 2016.

**DATA EXPLORATION: brand column**  
If some vehicle brands are not very common, they may be removed.

In [25]:
# View the unique brands and their frequency distribution.

brand_counts = autos['brand'].value_counts(normalize=True)
print(brand_counts.shape)
brand_counts.head(10)

(40,)


volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64

Most of the car brands are not very common. Only the brands that have a frequency greater than 5% will be retained.

In [26]:
# Create a list of the car brands that have a frequency > 5%.  
# Indices in brand_counts are car brands.  .index returns the car brands.

common_brands = brand_counts[brand_counts > 0.05].index
common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

**DATA ANALYSIS: Most expensive car brands.**

In [27]:
# Create a dictionary whose key:value pairs are brand:mean price.

brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand ] #filters dataframe for common brands
    mean_price = brand_only['price'].mean() #for a particular common brand, computes mean
    brand_mean_prices[brand] = int(mean_price) #adds key, values to dictionary

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Audi, BMW, Mercedes Benz are the most expensive vehicle brands. Ford and Opel are the least expensive, and Volkswagon is in the middle.

**DATA ANALYSIS: investigate if there is a correlation between the mean mileage and mean prices.**   

In [28]:
# Create mean mileage dictionary for top vehicle brands, where key:value pairs are brand:mean mileage.

brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [29]:
# Convert dictionaries, brand_mean_prices and brand_mean_mileage, to series.
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices)

# Convert one series to a dataframe.
brand_df = pd.DataFrame(mean_mileage, columns=['mean_mileage'])

# Assign the other series as a new column to the dataframe.
brand_df['mean_prices'] = mean_prices
brand_df

mean_mileage  mean_prices
bmw                  132572         8332
mercedes_benz        130788         8628
opel                 129310         2975
audi                 129157         9336
volkswagen           128707         5402
ford                 124266         3749

For the most common vehicle brands, the average odometer readings range from 124,000 km to 133,000 km, and is only about a 6% difference.  The price ranges from $3,749 to $8,332, which is a 55% difference. As the percent difference varies so greatly in the two variables, there does not appear to be a linear correlation betwen the mean mileage and mean prices.

## <font color='blue'>  CONCLUSION  </font>  

The purpose of this project was to clean the data and complete initial data analysis on the German eBay classified ads for used vehicles.

The following data cleaning was completed:
- Column header names: converted from camelcase to snakecase, and made names more descriptive  
- Price & odometer columns: removed non-numeric symbols, converted to integer type, and removed outliers  
- Registration_year column: removed dates prior to 1900  

Some analysis on the most common vehicle brands was completed:
- The most common brands were found to be: Volkswagen, BMW, Opel, Mercedes Benz, Audi.
- The most expensive brands were found to be: Audi, Mercedes Benz, BMW
- There was found to be no correlation between the mean mileage and mean prices of the most common vehicle brands.  

**Some potential next Steps to consider for future iterations of the project.**

Data Cleaning Next Steps:  
1. identify data that uses german words, translate them, map values to english  
2. convert dates to uniform integer data: 2016-03-21 -> 20160321  
3. any keywords in the name column that can be extracted as new columns

Analysis next steps:  
1. find most common brand/model combinations  
2. split the odometer_km into groups, use aggregation to see if average prices follow any patterns based on mileage  
3. how much cheaper are cars w/ damange and non-damaged counterparts